### PR TITLE
ADR: Switching to Deno Core

### PR DESCRIPTION
**Pull request by bajtos**

### CODE REVIEW

File: docs/architecture-decision-records/2023-01-switching-to-deno-core.md [hunk 1]
```diff
@@ -0,0 +1,126 @@
...
+
+While Deno is not perfect, we need to prioritize iteration speed at this stage
+of the project.
+
+## Consequences
+
+- We can support JS & Rust/WASM module from the beginning.
+
+- We can leverage what the Deno team learned about Rust & v8 integration over
+  the years and follow their best practices.
```
Comment (juliangruber) [inline docs/architecture-decision-records/2023-01-switching-to-deno-core.md:103]:
```suggestion
  the years and follow their best practices
```

🤓

File: docs/architecture-decision-records/2023-01-switching-to-deno-core.md [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+
+- We can cherry-pick functionality from Deno, either import some of their crates
+  or copy parts of their code (as a last resort solution).
```
Comment (juliangruber) [inline docs/architecture-decision-records/2023-01-switching-to-deno-core.md:106]:
```suggestion
  or copy parts of their code (as a last resort solution)
```

File: docs/architecture-decision-records/2023-01-switching-to-deno-core.md [hunk 3]
```diff
@@ -0,0 +1,126 @@
...
+
+### Caveats
+
+- Supporting both JS and Rust/WASM requires more work. Maybe we should pick one
+  as the first-class option with great documentation & DX, and keep the second
+  one less polished.
```
Comment (juliangruber) [inline docs/architecture-decision-records/2023-01-switching-to-deno-core.md:112]: Sgtm! Which are you leaning towards? Or can we create very minimal documentation, and get signals from early module creators?

**Review by juliangruber — CHANGES_REQUESTED** [review]
great work!

File: docs/architecture-decision-records/2023-01-switching-to-deno-core.md [hunk 4]
```diff
@@ -0,0 +1,126 @@
...
+### Raw v8
+
+v8 is a C++ project. Fortunately, the [Deno](https://deno.land) project
+maintains a [Rust wrapper](https://crates.io/crates/v8) converting v8's C++ APIs
+into APIs closer to idiomatic Rust.
+
+Unfortunately, working with v8 this way is still too cumbersome.
+
+See e.g. their example code here:
+[v8 example](https://docs.rs/v8/0.62.0/v8/#example)
```
[the remark this answers is not in the document]
Comment (bajtos) [reply docs/architecture-decision-records/2023-01-switching-to-deno-core.md:53]: Makes sense.

Part of the problem is that I don't even know where to find documentation on exposing custom functions using this crate.

I'll take another look.

File: docs/architecture-decision-records/2023-01-switching-to-deno-core.md (hunk 3 above)
Comment (bajtos) [reply docs/architecture-decision-records/2023-01-switching-to-deno-core.md:112]: I don't have a strong preference.

On second thought, if we find an easy way how to generate API docs automatically for both the JS & Rust SDKs, then the difference between "first-class" and "less-polished" can be pretty minimal

Comment (bajtos) [reply docs/architecture-decision-records/2023-01-switching-to-deno-core.md]: I decided to remove this sentence:

> Maybe we should pick one as the first-class option with great documentation & DX, and keep the second one less polished.

File: docs/architecture-decision-records/2023-01-switching-to-deno-core.md [hunk 5]
```diff
@@ -0,0 +1,126 @@
...
+add into a `deno_core`-based project.
+
+## Decision
+
+We decided to pivot and build Zinnia on top of `deno_core`.
+
+It will allow us to iterate much faster, while still having a lot of control of
+the API exposed to modules.
+
+While Deno is not perfect, we need to prioritize iteration speed at this stage
```
[the remark this answers is not in the document]
Comment (bajtos) [reply docs/architecture-decision-records/2023-01-switching-to-deno-core.md:95]: I added a link to the previous ADR where we are listing the issues. Is that good enough?

**Review by juliangruber — APPROVED** [review]

File: docs/architecture-decision-records/2023-01-switching-to-deno-core.md (hunk 4 above)
[the remark this answers is not in the document]
Comment (bajtos) [reply docs/architecture-decision-records/2023-01-switching-to-deno-core.md:53]: Uff!

I added more code to our `v8-js` PoC to show how to register a custom global function (e.g. a new Zinnia API) and implement it in Rust.

https://github.com/filecoin-station/runtime-poc/commit/337c4ffbc7d1264d775091e7494417aa551363e7

Do you still think it's not cumbersome?

I am afraid I have no idea how to connect `async fn` defined in Rust with v8 promises. I searched a bit in the `deno_core` codebase, but did not find anything that would help me to understand this better.

Here is the code that may be relevant (or where to start digging):

https://github.com/denoland/deno/blob/d318e38b76c8174d48fddfb99064401050cd8333/core/bindings.rs#L101-L251

Do you have any suggestions how to move forward with this?

**Pull request merged**